### PR TITLE
Feat: Context Parallel for Eagle3 Training

### PR DESCRIPTION
**Pull request by h-guo18**
## What does this PR do?

**Type of change:** New Feature 

**Overview:** 

- Supported Context Parallel by patching torch ring attention;
- Require following libirary version for stable cp: 
  - transformers5.0.0
  - accelrate1.12.0 
 - Move to FSDP2 
 - Removed unused arguments in training script (`--multi_gpu`, `fsdp_wrap_layer`)
 - Bump CI container to `nvcr.io/nvidia/pytorch:25.08-py3`

## Usage

```bash
./launch_train.sh --model $MODEL \
            --output_dir $OUTPUT_DIR \  
            --data $DATA \
            --num_epochs 0.1 \
            --train_bs 1 \
            --eagle_config eagle_config.json \
            --training_seq_len 1024 \
            --cp_size 2   #newly added
```

## Testing
- SDPA level correctness: tested TTT attention with/without CP, diff < 1%
```
=== Compare context-parallel (CP) outputs and grads with non-CP ===
Forward output comparison (CP vs Non-CP):
  Absolute diff (adiff) cp_out vs out: 0.001953125
  Relative diff (rdiff) cp_out vs out: 0.00182342529296875
WQ (query proj) grad comparison (CP vs Non-CP):
  Absolute diff (adiff) cp_wq_grad vs wq_grad: 0.0078125
  Relative diff (rdiff) cp_wq_grad vs wq_grad: 0.00347900390625
WK (key proj) grad comparison (CP vs Non-CP):
  Absolute diff (adiff) cp_wk_grad vs wk_grad: 0.0078125
  Relative diff (rdiff) cp_wk_grad vs wk_grad: 0.002471923828125
WV (value proj) grad comparison (CP vs Non-CP):
  Absolute diff (adiff) cp_wv_grad vs wv_grad: 0.25
  Relative diff (rdiff) cp_wv_grad vs wv_grad: 0.0069580078125
==============================================================
```

- E2E Training Acc
  (Llama3.1-8B, Unsynthesized magpie)
image

- Peak Mem Reserved
   (llama3.1-8B, 8xH100, train_length=4k)

    | cp_size | max_memory_allocated(MB) |max_memory_reserved (MB) |
    |----|--------------------------|--------------------------|
    | 1  |         65040.20              |79018.00
    | 2  |           50409.17             |73098.00
    | 4  |              45120.92            |72052.00
    | 8  |              38882.12            |66484.00

- Max Training Length test
  (llama3.1-8B, H100)

  | cp_size               | 6k  | 12k | 24k  | 48k  |
  |--------------------|-----|-----|-----|-----|
  | 1         | ✅ | OOM | OOM | OOM  |
  |2      | ✅  | ✅ | OOM | OOM |
  | 4         | ✅ | ✅ | ✅  | OOM  |
  | 8         | ✅ | ✅ | ✅  | ✅  |

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## Summary by CodeRabbit

* **New Features**
  * Added context parallelism (CP) and data parallelism shard size configuration parameters to training arguments.

* **Enhancements**
  * Improved TTT attention masking support for speculative decoding workflows.
  * Enhanced training launch script with improved parallelism configuration handling.

* **Chores**
  * Updated core dependencies: torch, transformers, accelerate, and wandb.
  * Added FSDP configuration file for distributed training setup.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: modelopt/torch/speculative/plugins/transformers.py [hunk 1]
```diff
@@ -56,10 +57,25 @@
...
+
+
+def _get_empty_cache(config):
+    """Return an empty cache. Handle different versions of transformers for unit tests."""
+    version = hf_transformers.__version__
+    major, minor = map(int, version.split(".")[:2])
+    if major > 4 or (major == 4 and minor > 53):
+        return DynamicCache(config=config)
+    else:
+        return DynamicCache()
```
Comment (kevalmorabia97) [inline modelopt/torch/speculative/plugins/transformers.py]: ModelOpt's supported transformers is `>=4.53.0`. So we only need special handling for `4.53.*`?

We can use following for better version check

```python
from packaging.version import Version

if Version(transformers.__version__) >= Version("4.54"):
    return DynamicCache(config=config)
else:
    return DynamicCache()
```

File: tests/examples/speculative_decoding/test_medusa.py [hunk 2]
```diff
@@ -32,7 +32,8 @@ def _run_hf_ptq(model_path, output_dir, qformat):
     )
 
 
-def test_llama_medusa_fp8_qat(tiny_llama_path, num_gpus, tiny_daring_anteater_path, tmp_path):
+@pytest.mark.skip(reason="speculative decoding uses transformers 5.x, quantization example uses transformers 4.x")
```
Comment (kevalmorabia97) [inline tests/examples/speculative_decoding/test_medusa.py]: We can still run the first medusa command and then add `pytest.skip` so still part of the test is run

File: modelopt/torch/speculative/utils.py [hunk 3]
```diff
@@ -27,8 +27,11 @@
 import torch.distributed
 from huggingface_hub import snapshot_download
 from torch import nn
+from torch.nn.attention import SDPBackend, sdpa_kernel
 from transformers.cache_utils import DynamicCache
 
+import modelopt.torch.speculative.plugins.transformers
```
Comment (kevalmorabia97) [inline modelopt/torch/speculative/utils.py]: You need to move this inside the `enable_cp_ttt_patch` function else it will fail if `transformers` is not available

File: modelopt/torch/speculative/plugins/transformers.py [hunk 4]
```diff
@@ -34,6 +34,7 @@
 from typing import Any
 
 import torch
+import transformers as hf_transformers
```
Comment (kevalmorabia97) [inline modelopt/torch/speculative/plugins/transformers.py]: Why do we import as `hf_transformers` here?

Comment (h-guo18) [reply modelopt/torch/speculative/plugins/transformers.py]: Otherwise modelopt.torch.speculative.plugins.transformers will point to the hf transformers library, rather than transformers.py, due to [this line.](https://github.com/NVIDIA/Model-Optimizer/blob/main/modelopt/torch/speculative/plugins/__init__.py#L33)

Comment (kevalmorabia97) [reply modelopt/torch/speculative/plugins/transformers.py]: You need to define `__all__ = [...]` to only include functions you want to use from this file otherwise python will import everything in this file

File: examples/speculative_decoding/requirements.txt [hunk 5]
```diff
@@ -1,5 +1,4 @@
-flash-attn
-openai
-py7zr
-sentencepiece>=0.2.0
-tensorboardX
+accelerate==1.12.0
+torch>=2.8.0,<2.9.0
+torchvision==0.23.0
```
Comment (kevalmorabia97) [inline examples/speculative_decoding/requirements.txt]: Lets not pin torch and torchvision in requirements else if user have different one and we overwrite it automatically, it may break other packages depending on the specific pre-installed torch.

Instead, we can check torch version in the example script and raise a warning with recommended version / docker image to run the example with

File: examples/speculative_decoding/README.md [hunk 6]
```diff
@@ -30,16 +30,21 @@ This example focuses on training with Hugging Face. To train with Megatron‑LM,
...
 ### Local Installation
 
-Install Modelopt with `hf` dependencies and other requirements for this example:
+To set up the environment locally, first install the latest ModelOpt with Hugging Face support:
+
+```bash
+pip install -e "../..[hf]"
+```
+
+Next, install any additional dependencies required for this example:
```
Comment (kevalmorabia97) [inline examples/speculative_decoding/README.md]: We dont need this since this change will anyway be in next release and users can continue doing pip install from stable version from pypi

File: examples/speculative_decoding/eagle_utils.py [hunk 7]
```diff
@@ -566,3 +576,142 @@ def on_step_end(self, args, state, control, **kwargs):
...
+
+    return patched_templated_attn
+
+
+def patch_ring_attention_for_ttt():
+    """Patch torch ring attention to support context parallelism for TTT."""
+    # Torch Ring Attention only supports no mask or causal mask. We apply the following patches to enable TTT mask.
+
+    if not (
+        Version(torch.__version__) > Version("2.7.0")
```
Comment (kevalmorabia97) [inline examples/speculative_decoding/eagle_utils.py]:
```suggestion
        Version(torch.__version__) > Version("2.7.1")
```

File: .github/workflows/example_tests.yml
Comment (kevalmorabia97) [inline .github/workflows/example_tests.yml]:
```suggestion
    needs: [check-file-changes, torch-pr, speculative-decoding-pr, trtllm-pr, onnx-pr]
```
